build_tools/packaging/promote_from_rc_to_final.py

```diff
@@ -0,0 +1,278 @@
+
+    print(" ...done")
+
+
+def promote_wheel(filename):
+    print(f"Promoting whl from rc to final: {filename}")
+
+    original_wheel = Wheel(filename)
+    original_version = Version(original_wheel.version)
+    base_version = original_version.base_version
+
+    print(f"  Detected version: {original_version}")
+
+    if original_version.local:  # torch packages
+        if not "rc" in original_version.local:
+            print("  Only release candidates (rc) can be promoted! Aborting!")
+            return False
+        local_version = str(original_version.local).split("rc", 1)[0]
+        base_version = original_version.public
+    else:  # rocm packages
+        if not "rc" in str(original_version):
+            print("  Only release candidates (rc) can be promoted! Aborting!")
+            return False
+        local_version = None
+
+    print(f"  New base version: {base_version}")
+    print(f"  New local version: {local_version}")
+
+    if str(base_version) == str(
+        original_version
+    ) or f"{base_version}+{local_version}" == str(original_version):
+        print("  Version is already a release version, skipping")
+        return False
+
+    print("  Starting to execute version change")
+    new_wheel_path = change_wheel_version.change_wheel_version(
+        pathlib.Path(filename),
+        str(base_version),
+        local_version,
+        callback_func=wheel_change_extra_files,
+    )
+    print("  Version change done")
+
+    new_wheel = Wheel(new_wheel_path)
+    new_version = Version(new_wheel.version)
+    print(f"New wheel has {new_version} and path is {new_wheel_path}")
+    return True
+
+
+def promote_targz(filename: str):
+    print(f"Found tar.gz: {filename}")
+
+    split = filename.removesuffix(".tar.gz").rsplit("/", 1)
+    base_dir = split[0]
+    dir_prefix = split[-1]
+    with tempfile.TemporaryDirectory(prefix=dir_prefix + "-") as tmp_dir:
+        print(f"  Extracting tar file to {tmp_dir}", end="")
+
+        targz = tarfile.open(filename)
+        targz.extractall(tmp_dir)
+        targz.close()
+        print(" ...done")
+
+        with open(f"{tmp_dir}/{dir_prefix}/PKG-INFO", "r") as info:
+            for line in info.readlines():
+                if "Version" in line:
+                    version = line.removeprefix("Version:").strip()
+
+        assert version, f"No version found in {filename}/PKG-INFO."
+
+        base_version = version.split("rc", 1)[0]
+
+        if any(c.isalpha() for c in base_version):
+            print(
+                f"  Base version extraction not successful and letters still in the version {base_version}."
+            )
+            print("  Only release candidates (rc) can be promoted! Aborting!")
+            return False
+        if base_version == version:
+            print(
+                f"  {version} and {base_version} are the same. Already the base version? Skipping..."
+            )
+            return False
```

Comment:
`packaging.version` provides `is_prerelease` to check if this is a released.

Reply:
i thought i could use it. but is_prerelease accepts way more than just "rc"

from their [doc](https://packaging.pypa.io/en/latest/version.html#packaging.version.Version.is_prerelease):
```
Version("1.2.3").is_prerelease
False
Version("1.2.3a1").is_prerelease
True
Version("1.2.3b1").is_prerelease
True
Version("1.2.3rc1").is_prerelease
True
Version("1.2.3dev1").is_prerelease
True
```

Reply:
so what would be the preferred choice?
check for "rc" only?
allow other prereleases for testing only?
always allow anything considered true with `is_prerelease`?

Reply:
Well the above checks if this is `is_prerelease` returns false or not?